src/zenml/integrations/vertex/orchestrators/vertex_orchestrator.py

```diff
@@ -0,0 +1,486 @@
+
+        image_name = self.get_docker_image_name(pipeline.name)
+
+        requirements = {*stack.requirements(), *pipeline.requirements}
+
+        logger.debug(
+            "Vertex AI Pipelines service docker container requirements %s",
+            requirements,
+        )
+
+        docker_utils.build_docker_image(
+            build_context_path=get_source_root_path(),
+            image_name=image_name,
+            dockerignore_path=pipeline.dockerignore_file,
+            requirements=requirements,
+            base_image=self.custom_docker_base_image_name,
+        )
+        container_registry.push_image(image_name)
+
+    def prepare_or_run_pipeline(
+        self,
+        sorted_steps: List["BaseStep"],
+        pipeline: "BasePipeline",
+        pb2_pipeline: "Pb2Pipeline",
+        stack: "Stack",
+        runtime_configuration: "RuntimeConfiguration",
+    ) -> Any:
+        """Creates a KFP JSON pipeline as intermediary representation of the pipeline
+        which is then deployed to Vertex AI Pipelines service.
+
+        How it works:
+        -------------
+        Before this method is called the `prepare_pipeline_deployment()` method
+        builds a Docker image that contains the code for the pipeline, all steps
+        the context around these files.
+
+        Based on this Docker image a callable is created which builds container_ops
+        for each step (`_construct_kfp_pipeline`). The function `kfp.components.load_component_from_text`
+        is used to create the `ContainerOp`, because using the `dsl.ContainerOp`
+        class directly is deprecated when using the Kubeflow SDK v2. The step
+        entrypoint command with the entrypoint arguments is the command that will
+        be executed by the container created using the previously created Docker
+        image.
+
+        This callable is then compiled into a JSON file that is used as the intermediary
+        representation of the Kubeflow pipeline.
+
+        This file then is submitted to the Vertex AI Pipelines service for execution.
+
+        Args:
+            sorted_steps: List of sorted steps.
+            pipeline: Zenml Pipeline instance.
+            pb2_pipeline: Protobuf Pipeline instance.
+            stack: The stack the pipeline was run on.
+            runtime_configuration: The Runtime configuration of the current run.
+
+        Raises:
+            ValueError: If the attribute `pipeline_root` is not set and it can be
+                not generated using the path of the artifact store in the stack
+                because it is not a `zenml.integrations.gcp.artifact_store.GCPArtifactStore`.
+        """
+
+        # If the `pipeline_root` has not been defined in the orchestrator configuration,
+        # try to create it from the artifact store if it is a `GCPArtifactStore`.
+        # Otherwise, raise an error and suggest the user to set it manually.
+        if not self.pipeline_root:
```

Comment:
Just one last question from my side: Does it make sense to have this logic in the `StackValidator` returned by the `validator` property? This validation is also called once a stack is registered, and we could warn users/prevent them from registering an invalid stack altogether.

Reply:
Ah... yes! I was doubting where to put this validation, but I didn't think the `validator` property which is the best place. Changing it!